server/src/main/java/org/apache/iotdb/db/query/udf/service/UDFRegistrationService.java

```diff
@@ -0,0 +1,320 @@
+      if (information.getClassName().equals(className)) {
+        String errorMessage;
+        if (information.isTemporary() == isTemporary) {
+          errorMessage = String
+              .format("UDF %s(%s) has already been registered successfully.",
+                  functionName, className);
+        } else {
+          errorMessage = String.format(
+              "Failed to register %sTEMPORARY UDF %s(%s), because a %sTEMPORARY UDF %s(%s) with the same function name and the class name has already been registered.",
+              isTemporary ? "" : "non-", functionName, className,
+              information.isTemporary() ? "" : "non-", information.getFunctionName(),
+              information.getClassName());
+        }
+        logger.warn(errorMessage);
+        throw new UDFRegistrationException(errorMessage);
+      } else {
+        String errorMessage = String.format(
+            "Failed to register UDF %s(%s), because a UDF %s(%s) with the same function name but a different class name has already been registered.",
+            functionName, className,
+            information.getFunctionName(), information.getClassName());
+        logger.warn(errorMessage);
+        throw new UDFRegistrationException(errorMessage);
+      }
+    }
+
+    Class<?> functionClass;
+    try {
+      udfClassLoader = getUDFClassLoader();
+      functionClass = Class.forName(className, true, udfClassLoader);
+      functionClass.getDeclaredConstructor().newInstance();
+    } catch (IOException | InstantiationException | InvocationTargetException | NoSuchMethodException | IllegalAccessException | ClassNotFoundException e) {
+      String errorMessage = String.format(
+          "Failed to register UDF %s(%s), because its instance can not be constructed successfully. Exception: %s",
+          functionName, className, e.toString());
+      logger.warn(errorMessage);
+      throw new UDFRegistrationException(errorMessage);
+    }
+
+    registrationInformation.put(functionName,
+        new UDFRegistrationInformation(functionName, className, functionClass, isTemporary));
+
+    if (writeToTemporaryLogFile && !isTemporary) {
+      try {
+        appendRegistrationLog(functionName, className);
+      } catch (IOException e) {
+        registrationInformation.remove(functionName);
+        String errorMessage = String
+            .format("Failed to append UDF log when registering UDF %s(%s), because %s",
+                functionName, className, e.toString());
+        logger.error(errorMessage);
+        throw new UDFRegistrationException(errorMessage, e);
+      }
+    }
+  }
+
+  private URLClassLoader getUDFClassLoader() throws IOException {
+    Collection<File> files = FileUtils
+        .listFiles(SystemFileFactory.INSTANCE.getFile(libRoot), null, true);
+    URL[] urls = FileUtils.toURLs(files.toArray(new File[0]));
+    return new URLClassLoader(urls);
+  }
+
+  public void deregister(String functionName) throws UDFRegistrationException {

```

Comment:
There may be a resource leak here.
I think you should unload the corresponding class in the URLClassLoader when the `deregister` function is called.